apbs/ainput/apbs_legacy_input.py

```diff
@@ -0,0 +1,1203 @@
+#       PRINT elecEnergy mol-1+mol-2-complex_mol END
+#
+#   The grammars and parsers convert all input and produce all output in
+#   lowercase alphanumeric representation. This could be a problem for
+#   filenames/pathnames.
+
+
+class ApbsLegacyInput:
+    """Class for reading in legacy APBS input files."""
+
+    def __init__(self):
+        """Setup parsing tokens and grammar for the APBS input file format."""
+
+        # Tokens used by multiple parsers
+        self.COMMENT = "#"
+        # TODO: The DEBUG flag should be set/accessed from a higher
+        #       level debugging object that is set from a config
+        #       file or command line option.
+        self.DEBUG = False
+        self.END_VAL = CLiteral("END")
+
+        # The FINAL_OUTPUT is a dictionary produced after parsing a file
+        # using the grammar and processing rules.
+        self.FINAL_OUTPUT = {}
+
+        # The highest level grammar to parse the READ, ELEC, APOLAR, and
+        # PRINT sections until the QUIT keyword is found.
+        self.grammar = OneOrMore(self.readParser()) + OneOrMore(
+            self.apolarParser() | self.elecParser()
+        ) + ZeroOrMore(self.printParser()) + Suppress(CLiteral("QUIT")) | (
+            # The following is one way to "narrow down" errors in an input
+            # file. It was adopted from:
+            # https://web.archive.org/web/20160821175151/http://pyparsing.wikispaces.com/share/view/30875955
+            # The idea is that the parsers above will succeed or the
+            # following will succeed and set the ParseSyntaxException
+            # stating that the token in the file is "unknown" and will
+            # show the line and column where the parser stopped. It does
+            # not always point out subtle errors like when a key is used
+            # in a section that does not support it. For example, using
+            # a keyword from pbam-auto in an ELEC section that specifies
+            # using tabi-auto type.
+            empty
+            - ~Word(printables).setName("<unknown>")
+        )
+
+    @staticmethod
+    def get_integer_grammar():
+        """Convenience function that is used by other classes."""
+        return Combine(Optional("-") + Word(nums))  # .setDebug()
+
+    @staticmethod
+    def get_identifier_grammar():
+        """Convenience function that is used by other classes."""
+        return (
+            Word(alphanums, alphanums + r"_" + r"-")
+            | ApbsLegacyInput.get_integer_grammar()
+        )  # .setDebug()
+
+    @staticmethod
+    def get_number_grammar():
+        """Convenience function that is used by other classes."""
+        return (
+            pyparsing_common.real | ApbsLegacyInput.get_integer_grammar()
+        )  # .setDebug()
+
+    @staticmethod
+    def get_path_grammar():
+        """Convenience function that is used by other classes."""
+        return Word(printables)  # .setDebug()
+
+    def debug(self, message: str):
```

Comment:
Why not use the `logging` module for this?

Reply:
I planned on using the logging module once we get more of the "pieces" implemented and part of an integration approach rather than a big bang approach. I can add it to this code or create an issue and add it later. I will let you make that call.

Reply:
NOTE: The `.setDebug()` method is part of pyparsing to print out matches. It has nothing to do with the `debug` method I wrote.

Reply:
I implemented using the `logging` module.

Reply:
Thanks -- I'd like to use `logging` as our starting point.  If we need interim debugging for development, we might want to use something super-explicit like https://pypi.org/project/icecream/